### DISCUSSION TITLE
Fallback for Postgresql DB in AsyncSession in Python

### DISCUSSION THREAD

**Discussion by junoriosity** [sqlalchemy/sqlalchemy#9064] [Usage Questions]
I am currently having two Postgresql Clusters attached to my Python FastAPI server

```
from sqlalchemy.ext.asyncio import AsyncSession, create_async_engine
from sqlalchemy.orm import sessionmaker

primary_engine = create_async_engine("PRIMARY_DATABASE_URL", echo=False, future=True)
replica_engine = create_async_engine("REPLICA_DATABASE_URL", echo=False, future=True)


async def get_primary_session() -> AsyncSession:
    async_session = sessionmaker(primary_engine, class_=AsyncSession, expire_on_commit=False)
    async with async_session() as session:
        yield session

async def get_replica_session() -> AsyncSession:
    async_session = sessionmaker(replica_engine, class_=AsyncSession, expire_on_commit=False)
    async with async_session() as session:
        yield session

```

Now assume, that the replica crashes. What would be the best strategy to redirect the traffic back to the primary?

I currently have tried this:

```
async def get_replica_session() -> AsyncSession:
    try:
        async_session = sessionmaker(replica_engine, class_=AsyncSession, expire_on_commit=False)
        async with async_session() as session:
            yield session
    except:
        async_session = sessionmaker(primary_engine, class_=AsyncSession, expire_on_commit=False)
        async with async_session() as session:
            yield session
```

However, it used to throw Runtime errors? Hence, it would be great if you could give me some thoughts about it.

**Comment by CaselIT** [comment]
Hi,

I don't think that will do what you seek. Also what does "the replica crashes" mean?

Btw you are using the sessionmaker wrongly. The only use it has is to create a module level object that can hold the session kws that can the be used tk create a session. If you create a new one each time you call the function you should instanciate the Session directly.
Does that pattern come from a particular documentation?

**Reply by junoriosity** [reply]
@CaselIT Many thanks for getting back to me.

I indeed got this from this tutorial [here](https://testdriven.io/blog/fastapi-sqlmodel/). In case, you have a better suggestion for using it in connection with FastAPI, please let me know, as this would be very helpful for me.

Also, in case you have a suggestion for my fallback issue, that would be very helpful. The idea is to have an idea of how to move the traffic to the primary, in case the replica crashes.

**Reply by CaselIT** [reply]
> I indeed got this from this tutorial [here](https://testdriven.io/blog/fastapi-sqlmodel/). In case, you have a better suggestion for using it in connection with FastAPI, please let me know, as this would be very helpful for me.

that bit in the tutorial seems not great. it does not make much sense creating a sessionmaker inside the function as it shown. just move it on the module, next to the create engine call.

regarding the rest of your question you would first need a way of detecting if the replica has crashed, since there may be a lot of exception raised by the db or driver that don't mean that the replica has crashed (since dbs are not really known for crashing so that would be a rare event). Once you have a way of detecting it I don't know how you can make fast api act on that. I don't think what you are trying works, since a single yield is used by fast api, but I'm not an expert on it at all, so you would have better luck asking in a fast api support forum.

**Reply by junoriosity** [reply]
@CaselIT could you give a short code snippet as example like you would do it. Then I might try it.

**Reply by CaselIT** [reply]
Sorry, I'm not sure what snippet are you asking for..

**Reply by junoriosity** [reply]
@CaselIT I mean the very first code snippet in my initial question. If you have some suggestions what I might try in order to improve it, I would be delighted.

**Reply by CaselIT** [reply]
as mentioned I don't think `get_replica_session` works correctly. I don't know how to do what you seek. I don't think it's possible to achieve what you want automatically.
As mentioned this is more of a fastapi question, not really a sqlalchemy one
